# 上机练习二
> 函数，递归，条件，循环


> 1-4 by Kaipeng Zeng, 5-8 By Meichao Luo, 9-12 By Luhua Jin, 13-16 By Yifei Zhang, 17-20 By Chuanming Liu, 21-25 By Weidong Wang, 26-29 By Yanming Liu, 30-33 By Sen Li, 34-37 By Xiaoyi Bao

## 1. Perfect Digital Invariant
### Description
An $N$ decimal number is called perfect digital invariant if and only if the sum of the $M$-th power of every digit is equal to itself. Now you are given $N$ and $M$ and you are supposed to find the smallest perfect digital invariant. (If there is no perfect digital invariant, return -1)

### Input
Two integers representing $N$ and $M$.

### Output
A integer representing the smallest perfect digital invariant.

### Note
$2\le N \le 6$.<br>
$1\le M \le 6$.

### Sample Input
6 5

### Sample Output
194979



In [2]:
# write your code here
# Any definitions and calls of other functions are allowed
def powerM(iN,M):
    strN=str(iN)
    p=0
    for i in strN:
        p+=int(i)**M
    return p
def get_perfect_digits(N, M):
    for i in range(10**(N-1),10**N):
        if powerM(i,M)==i:
            return i
    return -1

In [3]:
# run this block to verify the correctness of your code.
# Simply copy the answer is not allowed because TAs will review your codes.
testcase = [[3, 3], [4, 4], [4, 5], [5, 5], [6, 5], [6, 6]]
standard = [153, 1634, 4150, 54748, 194979, 548834]
for x in range(len(testcase)):
    case = testcase[x]
    std = standard[x]
    if get_perfect_digits(case[0], case[1]) == std:
        print("Correct")
    else:
        print("Wrong")

Correct
Correct
Correct
Correct
Correct
Correct


## 2. Unknown Result
### Description
In this part you are supposed to define a function $unknown(x, *args)$. Assume that the function accpet $N$ parameters in $*args$, we denote them as $p_0, p_2, \ldots, p_{N-1}$. The function need to return the value of $\sum_{i=0}^{N-1}(-1)^{i}\times p_{i}\times x^{i}$.
### Input 
A number $x$ and unknown number (not less than 2) of parameters.
### Output
The result of the equiation given in the description
### Sample Input
1 1 2
### Sample Output
-1
### Note 
Considering the errors that raise during the storage and calculation of float numbers, if the difference between your answer and the standard answer is not more than $10^{-3}$, you are correct.

In [5]:
# write your code here
def unknown(x, *args):
    p=0
    for i in range(len(args)):
        p+=(-1)**i*args[i]*x**i
    return p
unknown(1,1,2)

-1

In [6]:
# run this block to verify the correctness of your code.
# Simply copy the answer is not allowed because TAs will review your codes.
testcase = [
    [1, 1, 2], 
    [14, 7.68, 48.28, 50.32, 7.28], 
    [5, 6.92, 24.12, 20.36, 45.2, 10.88, 58.28], 
    [15, 32.96, 43.56, 17.88, 58.48, 22.12, 11.4], 
    [8, 27.36, 31.84, 39.08, 35.8, 46.96, 4.6, 46.2], 
    [10, 33.52, 38.08, 26.28], 
    [3, 37.44, 36.12, 14.68, 43.2, 10.28, 29.24]
]
standard = [-1, -10781.84, -180579.68, -7731017.4399999995, 12136612.32, 2280.7200000000003, -7377.84]
for t in range(len(testcase)):
    case = testcase[t]
    teststr = "unknown({}".format(case[0])
    for x in case[1:]:
        teststr += ',{}'.format(x)
    teststr += ")"
    if abs(eval(teststr) - standard[t]) < 1e-3:
        print('Correct')
    else:
        print("Wrong")

Correct
Correct
Correct
Correct
Correct
Correct
Correct


## 3. Palindrome
### Description
A palindrom string is a string that reads the same from left to right and from right to left. Now you are supposed to find the palindrome substring with maximum length. If there are many substrings, choose the first to appear.
### Input
A string $S$
### Output
$(x, y)$ representing two ends of the substring.
### Sample Input
"abbccssdss"
### Sample Output
(5, 9)

In [2]:
# write your code here
# any definitions and calls of other functions are allowed.
def palindrome_substring(S):
    length,left,right=0,0,0
    for i in range(len(S)):
        for j in range(i+1,len(S)):
            if S[i:j+1]==S[j+1:i:-1] and j+1-i>length:
                length,left,right=j+1-i,i,j+1
    return (left,right)
palindrome_substring('abbccssdss')

(5, 9)

In [3]:
# run this block to verify the correctness of your code.
# Simply copy the answer is not allowed because TAs will review your codes.
testcase = [
    'abbccssdss', 
    'ogivvssvvylcuke', 
    'jptxtmhhqcpojhiihjopcqbxcfuiqfanyfitb', 
    'kjthdhxvddvxhdnczb', 
    'aeyiigpazskoaburixuhfedbdrczgukzcyazzayczkuinblyfgfuhbbb', 
    'srxrcdxxtlalqvwckcjhmhfgkkqqkkgfhmiemcncpjktspypabsmpulm'
]
standard = [(5, 9), (3, 8), (8, 21), (4, 13), (29, 42), (20, 33)]
for x in range(len(testcase)):
    case = testcase[x]
    print("Correct" if palindrome_substring(case) == standard[x] else "Wrong")

Correct
Correct
Correct
Correct
Correct
Correct


## 4. Turning Man
### Description 
There is $n\times m$ matrix on the ground and there is a man at the position $(1,1)$. He is now going towards $(1,m)$ and he wants travel all over the matrix. every time he reachs a boundary of the matrix or he is going to step in a position he had already been to, he will turn right. every step he will only move to his nearby postion. Now TA ZKP wants to know where the man will be after $k$ steps ($k\le n\times m -1$) of moving, can you help him.
### Input
Three integers $n, m$ and $k$.
### Output
Two integers representing the position of the man.
### Sample Input
3 4 10
### Sample Output
2 2
### Explanation

In [59]:
# write your code here
def Move(N, M, k):
    pass

In [60]:
# run this block to verify the correctness of your code.
# Simply copy the answer is not allowed because TAs will review your codes.
testcase = [[3, 2, 2]]
standard = [(2, 2)]
for rg in range(len(testcase)):
    N, M, K = testcase[rg]
    print("Correct" if Move(N, M ,K) == standard[rg] else "Wrong")

Wrong


## 5. Geometric Progression 
### Description
Given two integers $a$, $b$ and an integer $p$, return all integers $m$ in the interval $[a,b]$ which starts from $a$ and every $m_{i+1} = m_i \times p$. 

### Input
An int $a$, an int $b$, an int $p$

### Output
Integers $m$
### Note
1. $\forall i, a\le m_i\le b$

## Example

### ex1.
Input: [10,50,2]

Output: 
<br>10
<br>20
<br>40



In [5]:
def geo_array(a,b,p):
    i=a
    res=''
    while i <=b:
        res=res+str(i)+'\n'
        i*=p
    return res

In [6]:
testcase=[[1,100,3],[100,2000,2],[1012,12381928394903810981908309131201,2321]]
for i in testcase:
    print(geo_array(i[0],i[1],i[-1]))

1
3
9
27
81

100
200
400
800
1600

1012
2348852
5451685492
12653362026932
29368453264509172
68164180026925788212
158209061842494754440052
367203232536430325055360692
852278702717054784453492166132



## 6. Geometric Progression Upgrade
### Description
Given two integers $a$, $b$ and an integer $p$, return all integers $m$ in the interval $[a,b]$ which starts from $a$ and every $m_{i+1} = m_i * p$ if i is odd and every $m_{i+1}=m_i* q$ if i is even.  

### Input
An int $a$, an int $b$, an int $p$, an int $q$

### Output
Integers $m$
### Note
1. $\forall i,a\le m_i\le b$

## Example

### ex1.
Input: [10,500,2,5]

Output: 
<br>10
<br>50
<br>100
<br>500

In [10]:
def geo_array_up(a,b,p,q):
    k=a
    res=''
    i=1
    while k<=b:
        res=res+str(k)+'\n'
        if i%2==0:
            k*=p
        else:
            k*=q
        i+=1
    return res

In [11]:
testcase=[[1,100,2,3],[20,200000000,1,13],[1000,12381928394903810981908309131201,1234,4321]]
for i in testcase:
    print(geo_array_up(i[0],i[1],i[-2],i[-1]))

1
3
6
18
36

20
260
260
3380
3380
43940
43940
571220
571220
7425860
7425860
96536180
96536180

1000
4321000
5332114000
23040064594000
28431439708996000
122852250982571716000
151599677712493497544000
655062207395684402887624000
808346763926274553163328016000



The output should be:
<br>1,
<br>3
<br>6
<br>18
<br>36

<br>20
<br>260
<br>260
<br>3380 <br>3380<br> 43940<br> 43940<br> 571220<br> 571220<br> 7425860<br> 7425860<br> 96536180<br> 96536180

<br>1000<br> 4321000
<br>5332114000<br> 23040064594000<br> 28431439708996000<br> 122852250982571716000<br> 151599677712493497544000<br> 655062207395684402887624000<br> 808346763926274553163328016000

## 7. Modular Inverse
  
### Description

Given a number $n$ and a mod $\phi (n<\phi)$, return a nubmer $m$ that $n\times m \equiv 1\mod \phi$. If no such number $m$, return $-1$


### Input 
An integer $n$, an integer $\phi$

### Output
An integer $m$


## Example

### ex1.
Input: [2,5]

Output: 3

       2*3 = 1 mod 5


### ex2.
Input: [2,4]

Output: -1

        gcd(2,4)!=1


In [1]:
import math
def get_inverse(n,phi):
    for i in range(1,phi+1):
        if (n*i-1)%phi==0:
            return i 
    return -1

In [2]:
testcase=[[7,15],[2,6],[917,1503]]
for i in testcase:
    print(get_inverse(i[0],i[-1]))

13
-1
890


The output should be:
<br>13.0
<br>-1
<br>890.0

## 8. Find Pattern
  
### Description

Given a string $n$ and a target pattern $target$ (also a string), print all int $m$ if there is a substring starts from index $m$ in string $n$ and this substring is the same as the pattern string $target$. 


### Input 
A string $n$, a string $target$

### Output
Integers 'm'

#### Note
1. The pattern may not show in the string
2. The length of target will be no longer than the string $(len(target)\le len(n))$

## Example
### ex1.
Input: ["adjajdsdadj","dj"]

Output: 
<br>1
<br>9

       The index of 'd' in 'dj'
### ex2.
Input: ["bbbbbbb","bbbbbb"]

Output: 
<br>0
<br>1

       multiple same patterns

In [5]:
def find_pattern(n:str,target:str) -> str:
    res=''
    for i in range(len(n)-len(target)+1):
        if n[i]==target[0]:
            for j in range(1,len(target)):
                if n[i+j]!=target[j]:
                    break
            res=res+str(i)+'\n'
    return res

In [6]:
testcase=[["abcdeabcdeabcdef","abc"],["abcd","efg"],["aaaaaaaa","aaa"]]
for i in testcase:
    print(find_pattern(i[0],i[-1]))

0
5
10


0
1
2
3
4
5



The output should be:
<br>0 <br> 5 <br> 10

<br>

<br>0<br> 1<br> 2<br> 3<br> 4<br> 5]

## 9. Simple Count
### Description
Given a sentence and calculate the number of letters and digits.

### Input
A string $s$

### Output
Two integers representing the number letters and digits

### Sample Input1
$s$ = "hello world! 123"

### Sample Output1
10 3

### Sample Input2
$s$ = "3.1415926"

### Sample Output2
0 8



In [12]:
# write your code here
# Any definitions and calls of other functions are allowed
def simple_count(s):
    letter,num=0,0
    for i in s:
        if i.isdigit():
            num+=1
        elif i.islower() or i.isupper():
            letter+=1
    return letter,num

(1, 1)

## 10. Greatest Common Divisor
### Description
Calculate the greatest common divisor of two positive integers.

### Input
Two positive integers $a$, $b$

### Output
The greatest common divisor

### Hint
Besides brute force, try Euclidean algorithm（辗转相除） to solve this problem

### Sample Input1
$a$ = 9, $b$ = 6

### Sample Output1
3

### Sample Input2
$a$ = 15750, $b$ = 27216

### Sample Output2
126



In [17]:
# write your code here
# Any definitions and calls of other functions are allowed
def gcd(a,b):
    a,b=min(a,b),max(a,b)
    while b%a!=0:
        a,b=b%a,a
    return a
print(gcd(15750,27216))
import math
print(math.gcd(15750,27216))

126
126


## 11. Valid Date
### Description
Given a string in the form of Year-Month-Day and judge whether it is a valid date.

### Input
a string in the form of Year-Month-Day

### Output
A boolen value

### Sample Input1
$date$ = "2004-2-29"

### Sample Output1
True

### Sample Input2
$date$ = "2003-2-29"

### Sample Output2
False

### Sample Input3
$date$ = "2004-2-30"

### Sample Output3
False

### Sample Input3
$date$ = "2004-3-30"

### Sample Output3
True

In [21]:
# write your code here
# Any definitions and calls of other functions are allowed
def is_leapyear(year:int):
    if year%400==0 or (year%100!=0 and year%4==0):
        return True
    else:return False
def date_valid(date):
    monthallow={1:31,3:31,5:31,7:31,8:31,10:31,12:31,4:30,6:30,9:30,11:30}
    item=[int(i) for i in date.split("-")]
    if item[1]==2:
        if is_leapyear(item[0]):
            if item[2]<=29 and item[2]>=1:
                return True
            else:return False
        else:
            if item[2]<=28 and item[2]>=1:
                return True
            else:return False
    else:
        if item[2]<=monthallow[item[1]]:
            return True
        else:return False

## 12. Recursive Multiplication
### Description
Use recursive to implement multiplication of two positive integers without using $*$. 

### Input
Two positive integers $A$, $B$

### Output
A positive integer

### Note
$+$, $-$, $<<$, $>>$ are allowed

### Sample Input1
$A$ = 1

$B$ = 10

### Sample Output1
10

### Sample Input2
$A$ = 3

$B$ = 4

### Sample Output2
12

In [25]:
# write your code here
# Any definitions and calls of other functions are allowed
def multiply(A,B):
    A,B=min(A,B),max(A,B)
    res=0
    for i in range(A):
        res+=B
    return res
x = 5
x = x << 1 # 二进制左移一位
print(x)

10


## 13. Find all digits
Given a string, find all digits in it.
### Input
A string s contains only lowercase letters and digits.
### Output
A string contains all digits in order of appearance in the given string s.
### Example
Input: '12fa35y6'

Output: '12356'

In [30]:
# write your code here
# Any definitions and calls to other functions are allowed
def findAllDigits(s):
    res=''
    for i in s:
        if i.isdigit():
            res+=i
    return res
def findAllDigits2(s):
    import re
    pattern1=re.compile("\d")
    return "".join(pattern1.findall(s))
print(findAllDigits('12fa35y6'))
print(findAllDigits2('12fa35y6'))

12356
12356


## 14. Product of all the factors
Given a integer $n$, find all its factors and the product of these numbers.
### Input
A integer $n$
### Output
A integer $m$
### Note
If $m > 10000$, output $m \mod 2333$
### Example
Input: 10

Output: 100(=1* 2* 5* 10)


In [34]:
# write your code here
# Any definitions and calls to other functions are allowed
def factorProduct(s1):
    if s1>10000:
        return f"{s1}  mod 2333"
    else:
        factors=[1,s1]
        for i in range(2,int(s1**0.5)+1):
            if s1%i==0:
                factors+=[i,s1/i]
        product=1
        for i in factors:
            product*=i
        factors=[str(int(item)) for item in factors]
        res=f"{product:.0f}(="+"*".join(factors)+")"
        return res
factorProduct(10)

'100(=1*10*2*5)'

## 15. Find all perfect number
Given a number $n$, find all perfect numbers which are strictly smaller than $n$.
### Input
A integer $n$
### Output
print all perfect number
### Note
A number $i$ is a perfect number if and only if $0 \le i$ and exist integer $k$ such that $k\times k-1=i$
### Example
Input: 4

Output: <br>
0<br>
3

In [35]:
# write your code here
# Any definitions and calls to other functions are allowed
def findPerfectNum(s1):
    p=1
    res=[]
    while p*p-1<s1:
        res.append(p*p-1)
        p+=1
    return res
findPerfectNum(100)

[0, 3, 8, 15, 24, 35, 48, 63, 80, 99]

## 16. Is Prime
Given two integer $m$ and $n$, decide if $num(m<num<n)$ is a prime. If $num$ is, print: 'num is a prime'. If $num$ is not a prime, print: 'num = f1 * f2', where f1,f2 is a pair of factors of $num$(any pair is OK).
### Input
two integer $m$, $n$
### Output
print output
### Example
Input: 9, 20

Output:

10 = 2 * 5

11 is a prime

12 = 2 * 6

13 is a prime

14 = 2 * 7

15 = 3 * 5

16 = 2 * 8

17 is a prime

18 = 2 * 9

19 is a prime

In [36]:
# write your code here
# Any definitions and calls to other functions are allowed
def factor(num:int)->tuple:
    if num==2:
        return 1,2
    if num%2==0:
        return 2,num//2
    for i in range(3,int(num**0.5)+1,2):
        if num%i==0:
            return i,num//i
    return 1,num
def isPrime(m,n):
    for i in range(m+1,n):
        if i==1:
            continue
        mi,ma=factor(i)
        if mi==1:
            print(f"{i} is a prime")
        else:print(f"{i}={mi}*{ma}")
    return
isPrime(9,20)        

10=2*5
11 is a prime
12=2*6
13 is a prime
14=2*7
15=3*5
16=2*8
17 is a prime
18=2*9
19 is a prime


## 17. Count trailing zeroes in factorial of a number

In [ ]:
def findTrailingZeros(n):
    p=0
    while n%10==0:
        n//=10
        p+=1
    return p
findTrailingZeros(100)

## 18. Design an algorithm to calculate Euler number (e). 
**Do not use any provided functions.**

100 digits precision is needed.

hints: check decimal usage in the first homework.

In [1]:
# write your code here


## 19. Is Subsequence
Given a string $s$ and a string $t$, check if $s$ is subsequence of $t$.

A subsequence of a string is a new string which is formed from the original string by deleting some (can be none) of the characters without disturbing the relative positions of the remaining characters. (ie, "ace" is a subsequence of "abcde" while "aec" is not).

### Example 1:

Input: s = "abc", t = "ahbgdc"\
Output: true<br>

### Example 2:

Input: s = "axc", t = "ahbgdc"\
Output: false
 

### Constraints:

$0 \le s.length \le 100$<br>
$0 \le t.length \le 10^4$\
Both strings consists only of lowercase characters.

In [43]:
def IsSubsequence(s:str,t:str)->bool:
    for letter in s:
        inde=t.find(letter)
        if inde==-1:
            return False
        t=t[0:inde]+t[inde+1:]
    return True

## 20. reverse a positive number
Give you a positive number, reverse this number,\
For example,\
input:123456\
output:654321
### Note
**Do not use any string related methods.**

In [ ]:
def reverseposnum(n:int)->int:
    p=[]
    while (n//10)!=0 or (n>0 and n<10):
        p.append(n%10)
        n=n//10
    dig10=len(p)
    q=0
    for i in range(dig10):
        q+=p[i]*10**(dig10-1-i)
    return q
print(reverseposnum(123456789))

## 21. Write the Result
Write the output of these programs without the help of computer. Then run these programs to check if your result is correct.

   ```python
   x = 1
   if x == 1:
       print("x=1")
       x = 2
   elif x == 2:
       print("x=2")
   else:
       print("x")
   print(x)
   ```

   ------------------
   
   ```python
   for i in range(1, 11, 2):
       print(i, end=' ')
       i = 9
   print()
   ```

Answer: 
x=1\
2\
1 3 5 7 9 

In [2]:
x = 1
if x == 1:
    print("x=1")
    x = 2
elif x == 2:
    print("x=2")
else:
    print("x")
print(x)
for i in range(1, 11, 2):
    print(i, end=' ')
    i = 9
print()

x=1
2
1 3 5 7 9 


## 22.
Given a non-negative integer $n$, write a function to do this: if $n < 10$, $n$ become the sum of all its digits. Repeat this process until $n < 10$. Return the final value.

```python
>>> print(add_digit(10))  # explanation: 1 + 0 = 1
1
>>> print(add_digit(555))  # explanation: 5 + 5 + 5 = 15; 1 + 5 = 6
6
```

In [5]:
def add_digit(digit:int)->int:
    if digit<10:
        return digit
    return add_digit(sum([int(i) for i in str(digit)]))
add_digit(555)

6

## 23.
Write function for an integer $n$. If $n$ is negative, you should let $n$ become its absolute value firstly. Then count the number of digit 1 in the binary form.

```python
>>> print(count_one(5))  # 5 is 0b101 in binary form (0b is the prefix to show that this number is in binary form)
2
>>> print(count_one(-7))  # Firstly -7 becomes 7. 7 is 0b111.
3
```

In [6]:
def count_one(i:int)->int:
    if i<0:
        i=-i
    item=list(bin(i))
    res=0
    for k in item:
        if k=='1':
            res+=1
    return res
count_one(7)

3

## 24.
Given a integer $n$, you have $k \ (k > 0)$ chances to guess this number. Use function "input()" to guess the number. If the number you guess is equal to $n$, print "Right!" and stop the function. If the number is smaller, print "Too small!". If the number is larger, print "Too large!". If you run out of the chances, print "The real number is " + str(n) + ".". Write a function (the first argument is $n$ and the second is $k$) to simulate it.

```python
>>> guess(9, 3)
Input: 2
Too small!
Input: 12
Too large!
Input: 11
Too large!
The real number is 9.
>>> guess(8, 3)
Input: 4
Too small!
Input: 8
Right!
```

In [9]:
def guess(n:int,k:int)->None:
    for i in range(k):
        x=int(input("Input: "))
        if x==n:
            print("Right!")
            return
        elif x>n:
            print("Too large!")
        elif x<n:
            print("Too small!")
    print(f"The real number is {n}.")
    return
guess(9,3)
guess(8,3)

Input:  2


Too small!


Input:  12


Too large!


Input:  11


Too large!
The real number is 9.


Input:  4


Too small!


Input:  8


Right!


## 25.
There are $n$ steps. Every time you can choose to climb 3 or 5 steps once. Calculate the number of different schemes to climb $n$ steps.

   ```python
   >>> print(climb(1))
   0
   >>> print(climb(2))
   0
   >>> print(climb(3))
   1
   >>> print(climb(15))  # Explanation: You can choose 3+3+3+3+3 or 5+5+5.
   2
   >>> print(climb(11))  # Explanation: You can choose 3+3+5 or 3+5+3 or 5+3+3.
   3
   ```
   
   Advanced Problem:  If you end up with less than five steps left, you can still use one jump to get to the top. Find the number of different schemes.
   
   ```python
   >>> print(climb_plus(1))  # Explanation: You can use 1 jump to reach the top.
   1
   >>> print(climb_plus(2))  # 2
   1
   >>> print(climb_plus(3))  # 3
   1
   >>> print(climb_plus(4))  # 4 or 3+1
   2
   >>> print(climb_plus(5))  # 5 or 3+2
   2
   >>> print(climb_plus(9))  # 3+3+3 3+5+1 5+4 5+3+1
   4
   >>> print(climb_plus(11))  # 3+3+3+2 3+3+5 3+5+3 5+3+3 5+5+1
   5
   >>> print(climb_plus(12))  # 3+3+3+3 3+3+5+1 3+5+3+1 3+5+4 5+3+3+1 5+3+4 5+5+2
   7
   ```

In [14]:
def climb(n:int)->int:
    if n==3 or n==5:
        return 1
    elif n in {1,2,4}:
        return 0
    elif n<=0:
        return "Error, n<=0"
    else:
        return climb(n-3)+climb(n-5)

def climb_plus(n:int)->int:
    if n in {1,2,3}:
        return 1
    elif n in {4,5}:return 2
    elif n>5:
        return climb_plus(n-3)+climb_plus(n-5)
    else: return "Error, unreasonable n."
    
for i in [1,2,3,4,5,9,11,12]:
    print(f"climb({i}):{climb(i)}")
    print(f"climb_plus({i}):{climb_plus(i)}")


climb(1):0
climb_plus(1):1
climb(2):0
climb_plus(2):1
climb(3):1
climb_plus(3):1
climb(4):0
climb_plus(4):2
climb(5):1
climb_plus(5):2
climb(9):1
climb_plus(9):4
climb(11):3
climb_plus(11):5
climb(12):1
climb_plus(12):7


## 26. Ring Target
### Description
Realize a function to print a $n\times n$ square ring target.<br/>
<img src="4-2.png" style="zoom:60%"/>
### Input
A positive integer $n$.
### Output 
A square ring target.
### Note
$1\leq n \leq 9$
### Example 1
#### Input
$n = 5$
#### Output
1 1 1 1 1 <br/>
1 2 2 2 1 <br/>
1 2 3 2 1 <br/>
1 2 2 2 1 <br/>
1 1 1 1 1
### Example 2
#### Input
$n = 8$
#### Output
1 1 1 1 1 1 1 1 <br/>
1 2 2 2 2 2 2 1 <br/>
1 2 3 3 3 3 2 1 <br/>
1 2 3 4 4 3 2 1 <br/>
1 2 3 4 4 3 2 1 <br/>
1 2 3 3 3 3 2 1 <br/>
1 2 2 2 2 2 2 1 <br/>
1 1 1 1 1 1 1 1

In [1]:
# write your code here
def Print_target(n):
    

In [2]:
for i in range(1, 10):
    Print_target(i)
    print()

## 27. Cutting line
### Description
Given a square $ABCD$, $A(-L,-L), B(L, -L), C(L, L), D(-L, L),\ L\in\mathbb{R}_+$, and a straight line $Ax+By+C=0$. Find the length of the segment cut by the square.
<img src="./4-1.png" style="zoom:50%"/>
### Input
A positive real number $L$. <br/>
Three real number $A,B,C$.
### Output
The length. <br/>*If the square and the straight line do not intersect, return 0.*
### Note
$0< L \leq 10$, $|A|,|B|,|C| < 10$. <br/>
**Accept floating point error ($<1e-4$).** <br/>
It is necessary to use some "if", but think twice before writing too many "if" to solve this problem.
### Example 1
#### Input
L = 4<br/>
A = 1<br/>
B = 1<br/>
C = -2
#### Output
8.485281374238571
### Example 2
#### Input
L = 2 <br/>
A = 1 <br/>
B = 0 <br/>
C = 1 
#### Output
4.0
### Example 3
#### Input
L = 3 <br/>
A = 3 <br/>
B = 1 <br/>
C = -3 
#### Output
6.324555320336759

In [6]:
# write your code here
def Cut_length(L, A, B ,C):
    return 0

In [7]:
import math
L = [4, 2, 3, 2, 1, 2, 5, 4, 4, 1, 5, 1, 3, 3, 5, 3, 4, 4]
A = [1, 1, 3, 2, 5, -3, 2, 0, 5, 4, 5, 0, 5, -2, 1, 0, 5, 2]
B = [1, 0, 1, 3, 3, 3, 5, -2, 0, -4, -5, -2, -3, 4, -4, -4, 1, 2]
C = [-2, 1, -3, 4, 1, 0, 7, 0, 15, 5, 25, 0, 17, 3, 18, 8, 13, 9]
Ans = [8.48528137423857, 4.0, 6.324555320336759, 3.6055512754639887, 2.33238075793812, 5.65685424949238, 10.770329614269007, 8.0, 8.0, 1.0606601717798216, 7.071067811865476, 2.0, 2.7211108842611424, 6.708203932499369, 7.2154348448309085, 6.0, 8.158431221748456, 4.949747468305833]
for i in range(len(L)):
    print(math.fabs(Cut_length(L[i], A[i], B[i], C[i]) - Ans[i]) < 1e-4)

False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False


## 28. Maximize Least Common Multiple
### Description
Given an interval $[L,R]\ (L,R\in\mathbb{N}_+, R - L \geq 2)$, choose $3$ integers from the interval to maximize their least common multiple.
### Input
Two integers $L$ and $R$.
### Output
The maximum of the least common multiple.
### Note
$R - L \geq 2$<br/> $1\leq L, R \leq 10^9$ <br/>
**hint**: <br/>$lcm(n, n+1) = n(n+1), \forall n\in\mathbb{N}_+$ <br/> Only a very small number of combinations are possible to be the answer<br/> To solve this problem, you may use "if".<br/>

### Example 1
#### Input
L = 2<br/>
R = 4
#### Output
12
### Example 2
#### Input
L = 3 <br/>
R = 8
#### Output
280
#### Explanation
lcm(5, 7, 8) = 280
### Example 3
#### Input
L = 4 <br/>
R = 9
#### Output
504
#### Explanation
lcm(7, 8, 9) = 504

In [25]:
# write your code here
def Max_lcm(L, R):
    pass

In [26]:
LR = [(7, 14), (5, 15), (5, 7), (3, 11), (8, 13), (99, 115), (25, 54), (5, 21), (50, 56), (62, 77), (410, 508), (472, 569), (357, 427), (319, 320), (281, 367),
(4087, 4417), (4804, 4884), (1416, 2114), (4133, 4670), (4211, 4593), (47911, 56531), (43506, 48130), (39014, 42050), (12507, 17061), (42951, 46772), 
(386090429, 429329748), (445568333, 502492199), (324960342, 343679806), (391138842, 467161404), (165489198, 247759951)]
Ans = [2002, 2730, 210, 990, 1716, 1481430, 140556, 7980, 163240, 438900, 130065780, 183249864, 77308350, 16230720, 49027530, 86116658880, 116357212086, 9429587902, 101760341410, 96829038096, 180649580020470, 111483729953790, 74345842441150, 4965203798940, 102310611552028, 79135789325242433717123190, 126878480555475364610374194, 40594017868129092236387490, 101953199503437070582535406, 15208742752825217952115050]
for i in range(len(LR)):
    print(Max_lcm(LR[i][0], LR[i][1]) == Ans[i])

False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False


## 29. Twin Primes
### Description
If $p$ and $p+2$ are both primes, $(p,p+2)$ are called twin primes. Given a positive number $n$, find the number of twin primes pair $(p,p+2)$ satisfying $p+2\leq n$.
### Input
A positive integer $n$
### Output
The number of twin primes pair
### Note 
$5\leq n\leq 1000$
### Example 1
#### Input
20
#### Output
4
#### Explanation
(3,5), (5,7), (11,13), (17,19)

In [10]:
# write your code here
def Twin_Prime(n):
    pass

## 30. Best Meeting Point
A group of two or more people wants to meet and minimize the total travel distance. You are given a 2D grid of values 0 or 1, where each 1 marks the home of someone in the group. The distance is calculated using Manhattan Distance, where distance$(p1, p2) = |p2.x - p1.x| + |p2.y - p1.y|$.

Example 1:
```
Input:
[[1,0,0,0,1],[0,0,0,0,0],[0,0,1,0,0]]
Output:
6

Explanation:
The point `(0,2)` is an ideal meeting point, as the total travel distance of `2 + 2 + 2 = 6` is minimal. So return `6`.
```
Example 2:
```
Input:
[[1,1,0,0,1],[1,0,1,0,0],[0,0,1,0,1]]
Output:
14
```

In [48]:
def minTotalDistance(grid):
    # Write your code here
    pass

In [50]:
print(minTotalDistance([[1,1,0,0,1],[1,0,1,0,0],[0,0,1,0,1]]))

None


## 31. Most Common Word
Given a paragraph and a list of banned words, return the most frequent word that is not in the list of banned words. It is guaranteed there is at least one word that isn't banned, and that the answer is unique.

Words in the list of banned words are given in lowercase, and free of punctuation. Words in the paragraph are not case sensitive. The answer is in lowercase.

Example1
```
Input:  paragraph = "Bob hit a ball, the hit BALL flew far after it was hit." and banned = ["hit"]
Output: "ball"
Explanation:
"hit" occurs 3 times, but it is a banned word.
"ball" occurs twice (and no other word does), so it is the most frequent non-banned word in the paragraph. 
Note that words in the paragraph are not case sensitive,
that punctuation is ignored (even if adjacent to words, such as "ball,"), 
and that "hit" isn't the answer even though it occurs more because it is banned.
```
Example2
```
Input:  paragraph = "a a a b b c c d" and banned = ["a","b"]
Output: "c"
Explanation:
"a" and "b" are banned words
"c" occurs 2 times and "d" occurs only once
So output "c"
```

In [31]:
def mostCommonWord(paragraph, banned):
    pass

In [32]:
para = "Bob hit a ball, the hit BALL flew far after it was hit."
banned = ["hit"]
print(mostCommonWord(para,banned))

None


## 32. Valid Triangle
Given three integers $a$, $b$, $c$, return true if they can form a triangle.

In [47]:
def isValidTriangle(a, b, c):
    # write your code here
    pass

## 33. Power of Two
Given an integer, write a function to determine if it is a power of two.

In [45]:
def isPowerOfTwo(self, n):
    # Write your code here
    pass


## 34. Representation with base 7

### Description
Given a positive integer $N$, return a string which is its octal representation.

The  numeral system of base 7 uses the digits 0 to 6. You may refer to [Base-10 to Base-7 Conversion Calculator](unitconversion.org/numbers/base-10-to-base-7-conversion.html) to check your answer.

### Input
0 < N < 100000, N is an integer

### Output
A string with each character from "0" to "6", and it does not contain any leading zeros.

### Note
Please try to implement it manually instead of using the built-in function for transformation in Python.

If you have no idea about how to do, you may read the definition of base 7 numeral system once again.

### SAMPLES
#### Example1
IN: 1

OUT: "1"


#### Example 2
IN: 9

OUT: "12"

#### Example3
IN: 100

OUT: "202"


#### Example4
IN: 2020

OUT: "5614"

#### Example5
IN: 100000

OUT: "564355"



## 35. Steps to reach 1

### Description
Given an integer number $n$. You are allowed to use the following operation to modify it.

- if $n$ is an even number, replace n with $\frac{n}{2}$
- if $n$ is an odd number, replace n with $n+1$

return the number of operations required to make $n$ exactly equal to $1$

### Input
An integer number $n$,  0 < $n$ < 10000

### Output
An integer number that represents the number of steps required to make it exactly 1.

### Samples

#### Example1

IN: 1

OUT: 0

#### Example2

IN: 2

OUT: 1

#### Example3

IN 3

OUT:3

#### Example4

IN: 17

OUT: 9


## 36. Number of inversion

### Description

You are given a list of numbers $nums$. Count the number of inversions in the list.

An inversion is defined as a tuple $(i, j)$ such that $nums[i] > nums[j]$.

### Input

A list of integers

$0 < len(nums) < 100$

### Output

An integer that represents the number of inversions

### Samples

#### Example1

IN: [1]

OUT: 0

#### Example2

IN: [1, 2, 3]

OUT: 0

#### Example3

IN: [2, 1, 3]

OUT: 1

#### Example4

IN: [4, 2, 1, 3]

OUT: 4

## 37. Buy Fruits

### Description

You have 100 Yuan to buy fruits and there are three kinds of fruits in the store you visit.  

The costs for fruits are:
- Banana:  1 for each
- Apple: 4 for each
- Orange: 5 for each

Return the number of different ways to spend exactly 100 Yuan to buy the three kinds of fruits. You need to buy each kind of fruits. For example, 0 banana, 5 apples, 16 oranges is not a valid solution.

### Input
None

### Output
An integer that represents the number of ways to spend exactly 100 Yuan with each kind of fruit bought.